add concurrency control for pipeline executors

serxa opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added concurrency control logic to limit total number of concurrent threads created by queries

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

serxa committed:
add concurrency control in pipeline executor

serxa commented:
low level approach to #36551

serxa committed:
fix comments

serxa commented:
## General description of ConcurrencyControl

Notion of cpu `slot` is introduced. Slot is unit of concurrency: to run a thread query has to acquire a slot in advance and release it when thread stops. Amount of slots is globally limited in a server. Multiple concurrent queries are competing for cpu slots if total demand exceed total amount of slots. `ConcurrencyControl` is responsible to resolve this competition by doing cpu slot scheduling in a fair manner. 

Each slot can be seen as independent state machine with the following states:
 * `free`: slot is available to be allocated by any query.
 * `granted`: slot is `allocated` by specific query, but not yet acquired by any thread.
 * `acquired`: slot is `allocated` by specific query and acquired by a thread.

Note that `allocated` slot can be in two different states: `granted` and `acquired`. The former is transitional state, that actually should be short. From instant when slot is allocated to a query till the moment when autoscaling procedure is run by any thread of a query (see example below).

```mermaid
stateDiagram-v2
    direction LR
    [*] --> free
    free --> allocated: allocate
    state allocated {
        direction LR
        [*] --> granted
        granted --> acquired: acquire
        acquired --> [*]
    }
    allocated --> free: release
```

API of `ConcurrencyControl` consists of the following functions:
1. **AllocationPtr ConcurrencyControl::allocate(UInt64 min, UInt64 max)**.
Create an allocation for a query: `auto slots = ConcurrencyControl::instance().allocate(min, max);`. It will allocate at least `min` and at most `max` slots. Note that `min` slots are granted immediately, but other `max - min` may be granted later. Thread-safe, non-blocking.
2. **SlotPtr Allocation::tryAcquire()**. For every thread a slot has to be acquired from an allocation, e.g.:`while (auto slot = slots->tryAcquire()) createYourThread([slot = std::move(slot)] { ... });`. This snippet can be used at query startup and for upscaling later. Thread-safe, non-blocking.
3. **ConcurrencyControl::setMaxConcurrency(limit)**. Update the total amount of slots. Thread-safe, non-blocking.

## Integration of ConcurrencyControl with PipelineExecutor
Example below shows lifespan of a single query. Startup with one thread and later autoscaling (green) to two threads. Note that worker threads have to periodically check if there are granted slots because there are no notifications.

```mermaid
sequenceDiagram
    participant CC as Concurrency Control
    participant Q as Query
    participant T as Thread
    participant T2 as Thread

    Q->>CC: allocate(1, max_threads)
    CC->>Q: allocation(granted:1)
    Note over Q: keeps allocation
    Note over Q: aquire granted slot(s)
    Q->>T: spawn_thread(slot)
    activate T

    rect rgb(230, 255, 200)
   
    Note over CC: someone releases a slot
    CC->>Q: grant(slot) to allocation
    T->>Q: allocation.tryAcquire()
    Q->>T: acquired slot
    T->>T2: spawn_thread(slot)
    end
    activate T2

    T->>Q: release(slot)
    deactivate T

    T2->>Q: release(slot)
    deactivate T2

    Note over Q,T2: join threads

    Q->>CC: destroy(allocation)
```

## ConcurrencyControl implementation details
Main points of implementation:
 * ConcurrencyControl is internally represented by two counters (currently allocated slots and total amount of slots) plus list of waiting allocations.
 * Released slots are distributed between waiting allocations in a round-robin manner to provide some kind of fairness (but actually implementation cannot be fair without downscaling).
 * Oversubscription is possible: total amount of allocated slots can exceed `setMaxConcurrency(limit)`, because `min` amount of slots is allocated for each query unconditionally.

ilejn commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,207 @@
+    using Waiters = std::list<Allocation *>;
+
+    // Scoped guard for acquired slot, see Allocation::tryAcquire()
+    struct Slot : boost::noncopyable
```
Slot looks like a std::shared_ptr<Allocation, Deleter> with custom deleter.
It is how a scope guard should (or at least can) be implemented in modern C++
Not sure if an explicit class makes things more verbose or vice versa.

serxa replied:
No, it's actually destructed independent from Allocation destruction

ilejn commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,207 @@
+                return;
+
+            std::unique_lock lock{parent.mutex};
+            parent.cur_concurrency -= allocated - released;
```
Is it really needed to use parent.mutex directly?
Why not make ConcurrencyControl interface complete?

serxa replied:
Just to keep things as simple as possible, they are tightly bound anyways now. Probably can be improved later if someone needs more complicated logic without Allocation.

ilejn replied:
Thanks for handling it

ilejn commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,207 @@
+public:
+    struct Allocation;
+    using AllocationPtr = std::shared_ptr<Allocation>;
+    using Slots = UInt64;
```
To me Slots is a vector of Slot
NSlots may be?

serxa replied:
In this implementation I only create SlotPtr object when slot is allocated, just to hold its ownership. Probably Slots is not the best name, maybe SlotCount is better?

ilejn commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,207 @@
+    std::mutex mutex;
+    Waiters waiters;
+    Waiters::iterator cur_waiter; // round-robin pointer
+    Slots max_concurrency = Slots(-1);
```
std::numeric_limits<Slots>::max  ?
Alternatively, std::thread ::hardware_concurrency() , while it is rather problematic option, e.g. it can return 0.

ilejn commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,207 @@
+            if (allocation->grant())
+                ++cur_waiter;
+            else
+                waiters.erase(cur_waiter++); // last required slot has just been granted -- stop waiting
```
To me cur_waiter = waiters.erase(cur_waiter) is safer and more generic, although postfix increment probably fine for list.

serxa committed:
add tests; fix bugs; make Allocation and ConcurrencyControl classes less tighly bounded

ilejn commented:
Looks great to me!

serxa commented:
@mergifyio update

mergify[bot] committed:
Merge branch 'master' into concurrency-control

serxa committed:
fix typos

serxa committed:
Merge branch 'concurrency-control' of github.com:ClickHouse/ClickHouse into concurrency-control

KochetovNicolai commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,234 @@
+            if (!granted)
+                return {};
+            granted--;
+            return SlotPtr(new Slot(shared_from_this()));
```
make_shared?

serxa replied:
Constructor is private to avoid misuse

KochetovNicolai replied:
ah, now I see.
Could you add a comment? :)

KochetovNicolai commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,234 @@
+        // Create allocation and start waiting if more slots are required
+        auto allocation = new Allocation(*this, limit, granted);
+        if (granted < limit)
+            allocation->wait(waiters.insert(cur_waiter, allocation));
```
I think it would be better to set iterator into Allocation ctor.
First, you can insert nullptr into waiters, then replace value to `allocation.get()`

serxa commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,234 @@
+        cur_concurrency += granted;
+
+        // Create allocation and start waiting if more slots are required
+        auto allocation = new Allocation(*this, limit, granted);
```
Also just because constructor is private. Using raw pointer here should be safe (unless `std::list::insert()` throws, in which case we anyways have big troubles), and it looked to me a little bit cleaner. But I can change it to shared_ptr if it's not very convincing.

KochetovNicolai commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,234 @@
+    {
+        ~Allocation()
+        {
+            parent.free(this); // We have to lock parent's mutex to avoid race with grant()
```
Can we check some invariants here? maybe `granted = 0 && allocated == released`.

Also, in comment you said that life cycle of slot is `granted -> acquired -> free`.
So maybe we can have counters `granted`, `acquired`, `freed` which would tell how many slots are passed this state. There would be global invariant `granted >= acquired >= freed` and all the counters would never decrease.

serxa replied:
Oh, it actually turned out that granted also can be non-zero, so everything is possible - no checks

serxa commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,234 @@
+    void free(Allocation * allocation)
+    {
+        std::unique_lock lock{mutex};
+        auto [amount, waiter] = allocation->free();
```
Yes, unfortunately it is important to avoid race, `grant()` can happen into the same allocation when we destroy allocation which is still waiting for more slots (this is in fact canceling, that was hard to be done in previous design of upscaling that we discussed with you).

Locking parent's mutex prevents slot scheduling and thus `grant()` to be called. Also in earlier implementation these two classes `ConcurrencyControl` and `Allocation` were more tightly coupled and there was weird `unique_lock lock{parent.mutex}` construction, but @ilejn convinced me to remove it.

KochetovNicolai replied:
Ok.
Another option is, probably, add some flag which tells that Allocation is cencelled.
And grant can check that flag and return some special status that it was not successful.
Anyway, a comment will be useful :)

serxa commented on src/Processors/Executors/PipelineExecutor.cpp:
```diff
@@ -265,6 +267,56 @@ void PipelineExecutor::initializeExecution(size_t num_threads)
+
+void PipelineExecutor::spawnThreads()
+{
+    while (auto slot = slots->tryAcquire())
```
If we don't have granted slots tryAcquire will return `{}`, i.e. `nullptr` containing `SlotPtr` and it will be interpreted as `false`.

Yeah, there is no condvar. I decided that periodic checking is okay, because:
 - it was hard to implement waiting on condvar: it requires some thread to be waiting
 - we anyway have at least one active thread that can do try to do upscaling by calling `spawnThreads()`
 - I expect overhead of periodic `tryAcquire()` calls to be low.

KochetovNicolai commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,234 @@
+        std::unique_lock lock{mutex};
+
+        // Acquire as much slots as we can, but not lower than `min`
+        SlotCount limit = std::max(min, max);
```
maybe check that `min <= max`

KochetovNicolai reviewed (approved):
Generally LGTM

KochetovNicolai commented on src/Common/ConcurrencyControl.h:
```diff
@@ -0,0 +1,234 @@
+    };
+
+    // FIXME: have to be unique_ptr, but ThreadFromGlobalPool does not support move semantics yet
+    using SlotPtr = std::shared_ptr<Slot>;
```
ah, sorry

serxa commented on src/Processors/Executors/PipelineExecutor.cpp:
```diff
@@ -240,6 +239,9 @@ void PipelineExecutor::executeStepImpl(size_t thread_num, std::atomic_bool * yie
             }
 
+            // Upscale if possible
+            spawnThreads();
```
Ok!

serxa replied:
After implementing lock-free tryAcquire it looks unnecessary and also complicated because `num_threads` is not available in this function (requires either a new member `num_threads` or changing signature of couple private functions)

KochetovNicolai commented on src/Processors/Executors/PipelineExecutor.cpp:
```diff
@@ -265,6 +267,56 @@ void PipelineExecutor::initializeExecution(size_t num_threads)
     tasks.init(num_threads, profile_processors);
     tasks.fill(queue);
+
+    slots = ConcurrencyControl::instance().allocate(1, num_threads);
```
> You think it is not good enough?

I just think that (maybe) it is good to have an ability to disable a new feature by default :)
Less codepaths is also a good reason not to do it.

serxa committed:
perf/safety/docs improvements

antonio2368 commented:
What about moving implementation details to cpp?
What I'm interested the most is if you can hide the Slot definition completely that way because you return pointers to it.

rvasin commented:
@serxa This is a short message to let you know. I built your branch as it is. It works fine.
Then I did:
```
git fetch origin
git merge origin/master
git submodule update --init --recursive
```
After your branch merged with upstream and re-building it produces many Fatal error messages. This is for the case when config.xml has query_log section:
```
    <query_log>
        <database>system</database>
        <table>query_log</table>
    </query_log>
```
Then I start CH with command:
`./clickhouse-server --config-file /etc/clickhouse-server/config.xml`

And I have the error messages:

> 
> 2022.06.03 10:21:18.848742 [ 53904 ] {046516c9-25b8-4a53-9bba-08236ace2748::202205_12333_19587_1094} <Debug> MergeTreeSequentialSource: Reading 118 marks from part 202205_14827_16619_1040, total 958072 rows starting from the beginning of the part
> 2022.06.03 10:21:18.848759 [ 53903 ] {046516c9-25b8-4a53-9bba-08236ace2748::202205_19979_19981_1} <Debug> MergeTreeSequentialSource: Reading 2 marks from part 202205_19981_19981_0, total 20 rows starting from the beginning of the part
> 2022.06.03 10:21:18.848767 [ 53909 ] {046516c9-25b8-4a53-9bba-08236ace2748::202204_7897_12328_1254} <Debug> MergeTreeSequentialSource: Reading 15 marks from part 202204_11274_12328_908, total 112984 rows starting from the beginning of the part
> 2022.06.03 10:21:18.848754 [ 54099 ] {} <Fatal> BaseDaemon: ########################################
> 2022.06.03 10:21:18.848830 [ 54099 ] {} <Fatal> BaseDaemon: (version 22.6.1.1, build id: 5A59244E0B2943E8) (from thread 53900) (query_id: b41db5ab-72b4-4213-8b64-0b701be65f99::20220502_1098_1166_9) (query: ) Received signal Segmentation fault (11)
> 2022.06.03 10:21:18.848851 [ 54099 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.
> 2022.06.03 10:21:18.848866 [ 53912 ] {c4128e59-3e2c-4563-8bb5-dfd50bcad3f2::202205_1192_1389_40} <Debug> MergeTreeSequentialSource: Reading 2 marks from part 202205_1387_1387_0, total 82 rows starting from the beginning of the part
> 2022.06.03 10:21:18.848883 [ 53904 ] {046516c9-25b8-4a53-9bba-08236ace2748::202205_12333_19587_1094} <Debug> MergeTreeSequentialSource: Reading 70 marks from part 202205_16620_19587_1093, total 558058 rows starting from the beginning of the part
> 2022.06.03 10:21:18.848878 [ 53903 ] {046516c9-25b8-4a53-9bba-08236ace2748::202205_19979_19981_1} <Trace> SortDescription: Compile sort description Date 1 1DateTime 1 1
> 2022.06.03 10:21:18.848885 [ 54099 ] {} <Fatal> BaseDaemon: Stack trace: 0x14a890bc 0x14a8ba28 0x14a8bca7 0x1381ca3f 0x14a8aec7 0x14a89ec5 0x14a8d6b8 0x14cba035 0x14cb9d3a 0x14e9e616 0x14cd20da 0x14cc898a 0x14cc85a0 0x14cd57d8 0x14cd5a4e 0x14969514 0x1496946b 0x1496cd1a 0x14b81ced 0x1497a30b 0x1497b099 0xb08a1e9 0xb08bdad 0xb08899a 0xb08acae 0x7f7849f2b609 0x7f7849e50133
> 2022.06.03 10:21:18.849215 [ 53912 ] {c4128e59-3e2c-4563-8bb5-dfd50bcad3f2::202205_1192_1389_40} <Debug> MergeTreeSequentialSource: Reading 2 marks from part 202205_1388_1388_0, total 172 rows starting from the beginning of the part
> 2022.06.03 10:21:18.849483 [ 53912 ] {c4128e59-3e2c-4563-8bb5-dfd50bcad3f2::202205_1192_1389_40} <Debug> MergeTreeSequentialSource: Reading 2 marks from part 202205_1389_1389_0, total 146 rows starting from the beginning of the part
> 2022.06.03 10:21:18.850620 [ 53982 ] {} <Warning> system.metric_log_0 (0aa81d8d-2e85-4db8-9fdb-8f7bce36aa3a): Removing temporary directory /var/lib/clickhouse/store/0aa/0aa81d8d-2e85-4db8-9fdb-8f7bce36aa3a/tmp_merge_202205_19278_32619_1037/
> 2022.06.03 10:21:18.855149 [ 54099 ] {} <Fatal> BaseDaemon: 3. ./build/../src/Storages/MergeTree/MergeTreeReaderStream.cpp:267: DB::MergeTreeReaderStream::adjustRightMark(unsigned long) @ 0x14a890bc in /home/rvasin/cpp/sergeibranchlatest/ClickHouse/build/programs/clickhouse
> 2022.06.03 10:21:18.863910 [ 53905 ] {68c6f587-50f9-4fad-af21-1be46f795685::20220422_7_9_2} <Debug> system.text_log (68c6f587-50f9-4fad-af21-1be46f795685) (TTLTransform): Removed all rows from part 20220422_7_9_2 due to expired TTL
> 2022.06.03 10:21:18.864052 [ 53890 ] {} <Trace> BaseDaemon: Received signal 6
> 2022.06.03 10:21:18.864149 [ 53890 ] {} <Trace> BaseDaemon: Received signal 11
> 2022.06.03 10:21:18.864231 [ 54100 ] {} <Fatal> BaseDaemon: ########################################
> 2022.06.03 10:21:18.864306 [ 54100 ] {} <Fatal> BaseDaemon: (version 22.6.1.1, build id: 5A59244E0B2943E8) (from thread 53905) (query_id: 68c6f587-50f9-4fad-af21-1be46f795685::20220422_7_9_2) (query: ) Received signal Aborted (6)
> 2022.06.03 10:21:18.864334 [ 54100 ] {} <Fatal> BaseDaemon: 
> 2022.06.03 10:21:18.864305 [ 54101 ] {} <Fatal> BaseDaemon: ########################################
> 2022.06.03 10:21:18.864363 [ 54100 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f7849d7400b 0x7f7849d53859 0xb08a670 0x14cc7fec 0x14cd565e 0x14969656 0x1496946b 0x1496cd1a 0x14b81ced 0x1497a30b 0x1497b099 0xb08a1e9 0xb08bdad 0xb08899a 0xb08acae 0x7f7849f2b609 0x7f7849e50133
> 2022.06.03 10:21:18.864379 [ 54101 ] {} <Fatal> BaseDaemon: (version 22.6.1.1, build id: 5A59244E0B2943E8) (from thread 54098) (no query) Received signal Segmentation fault (11)
> 2022.06.03 10:21:18.864392 [ 54100 ] {} <Fatal> BaseDaemon: 3. gsignal @ 0x7f7849d7400b in ?
> 2022.06.03 10:21:18.864400 [ 54101 ] {} <Fatal> BaseDaemon: Address: 0xe0 Access: write. Address not mapped to object.
> 2022.06.03 10:21:18.864406 [ 54100 ] {} <Fatal> BaseDaemon: 4. abort @ 0x7f7849d53859 in ?
> 2022.06.03 10:21:18.864419 [ 54101 ] {} <Fatal> BaseDaemon: Stack trace: 0x14cd20e1 0x14cc898a 0x14cc9ecf 0xb08899a 0xb08acae 0x7f7849f2b609 0x7f7849e50133
> 2022.06.03 10:21:18.865825 [ 54099 ] {} <Fatal> BaseDaemon: 4. ./build/../src/Storages/MergeTree/MergeTreeReaderWide.cpp:219: DB::getStream(bool, DB::ISerialization::SubstreamPath const&, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::unique_ptr<DB::MergeTreeReaderStream, std::__1::default_delete<DB::MergeTreeReaderStream> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::unique_ptr<DB::MergeTreeReaderStream, std::__1::default_delete<DB::MergeTreeReaderStream> > > > >&, DB::NameAndTypePair const&, unsigned long, bool, unsigned long, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > >&) @ 0x14a8ba28 in /home/rvasin/cpp/sergeibranchlatest/ClickHouse/build/programs/clickhouse
> 2022.06.03 10:21:18.867049 [ 54101 ] {} <Fatal> BaseDaemon: 3. ./build/../src/Processors/Executors/ExecutionThreadContext.cpp:70: DB::ExecutionThreadContext::executeTask() @ 0x14cd20e1 in /home/rvasin/cpp/sergeibranchlatest/ClickHouse/build/programs/clickhouse
> 2022.06.03 10:21:18.869967 [ 54100 ] {} <Fatal> BaseDaemon: 5.1. inlined from ./build/../contrib/libcxx/include/__memory/shared_ptr.h:174: std::__1::__shared_count::__release_shared()
> 2022.06.03 10:21:18.869991 [ 54100 ] {} <Fatal> BaseDaemon: 5.2. inlined from ../contrib/libcxx/include/__memory/shared_ptr.h:216: std::__1::__shared_weak_count::__release_shared()
> 2022.06.03 10:21:18.870005 [ 54100 ] {} <Fatal> BaseDaemon: 5.3. inlined from ../contrib/libcxx/include/__memory/shared_ptr.h:703: ~shared_ptr
> 2022.06.03 10:21:18.870017 [ 54100 ] {} <Fatal> BaseDaemon: 5. ../src/Common/ThreadPool.h:210: ThreadFromGlobalPool::~ThreadFromGlobalPool() @ 0xb08a670 in /home/rvasin/cpp/sergeibranchlatest/ClickHouse/build/programs/clickhouse
> 2022.06.03 10:21:18.874392 [ 54101 ] {} <Fatal> BaseDaemon: 4. ./build/../src/Processors/Executors/PipelineExecutor.cpp:216: DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x14cc898a in /home/rvasin/cpp/sergeibranchlatest/ClickHouse/build/programs/clickhouse

Please test your branch with upstream master changes.

serxa commented:
Thanks for trying it out! I'll investigate this more closely during next week.

rvasin committed:
Fix Fatal errors caused by using spawnThreads()

rvasin commented:
@serxa Good news: I corrected the issue with Fatal error messages. Please read the [message](https://github.com/ClickHouse/ClickHouse/pull/37285#issuecomment-1159002168) in my pull request.

serxa commented:
For now this implementation got issues with concurrent access to `ExecutionThreadContext::node` field. Thread sanitizer output is [here](https://pastila.nl/?006c8193/63afca52718b653b5996f7dd0023fddf). I'm investigating it...

serxa committed:
fix single-thread mode for pulling and pushing executors to work properly

serxa commented:
Problem was in the separate "single-threaded" mode used by pulling and pushing executors. They never create a thread, but do all logic inside external (outside PipelineExecutor) thread using `executeStep()`. So there was no logic that acquires a slot for it, so on `spawnThreads()` inside `executeStep()` it figured out that there is 0/1 slot - let's upscale! After that we had 2 thread executing with `thread_num == 0` (one external and one spawned). That caused a lot of races. Fixed now.
@rvasin Here is the fix, all thread-sanitizer errors I had are gone now.

serxa committed:
unittest style/perf minor fixes

mergify[bot] committed:
Merge branch 'master' into concurrency-control

serxa commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into concurrency-control

serxa committed:
Merge branch 'master' into concurrency-control

serxa committed:
fix finish() condition to account only active threads

serxa commented:
Problem with hanging session fixed.

This hang up happened due to query that is implicitly done by client at start. That query uses high `max_threads` and if it happens to be higher than `total_max_threads` it hangs. Root cause was improper condition for PipelineExecution finish() call. It checked that there are no tasks and all `num_threads` are waiting, but in our case not all thread were actually started in the first place, so condition was never evaluated to true. I've added tracking number of started threads (aka `use_threads`) and fixed that condition.

@rvasin Now I have no more hangups, could you please pull changes into your branch and test it a little bit?

serxa commented:
I'll leave patch for debugging [here](https://pastila.nl/?00004ca7/82a7eaae35561c59addab6cbb9cc0a3b) just in case

rvasin commented:
> @rvasin Now I have no more hangups, could you please pull changes into your branch and test it a little bit? UPD. Probably better to wait tests here to finish successfully

@serxa Sergei thank you for the update. I will integrate tests to my branch and I will do some tests later on today.

BTW I could not be sure. But in some situation few days ago the errors were gone in case of disabling the profiler in users.xml this way:
```

    <profiles>                                                                                             
        <default>                                                                                          
            <log_queries>1</log_queries>                                                                                                                  
            <max_memory_usage>10000000000</max_memory_usage>                                               
            <use_uncompressed_cache>0</use_uncompressed_cache>                                             
            <!-- disable profiler: -->                                                                     
            <query_profiler_cpu_time_period_ns>0</query_profiler_cpu_time_period_ns>                       
            <query_profiler_real_time_period_ns>0</query_profiler_real_time_period_ns>                     
        </default>       
                                                                       
    </profiles> 
```           

I did not have time to look how the profiler is invoked. Do you remember we had similar problem with Flush Logs thread and you fixed it?  So I will do some tests of profiler options too and analyze the source code.

serxa commented:
> Do you remember we had similar problem with Flush Logs thread and you fixed it?

No, my fix only for new code added in this PR, not related to logs. But iirc there was deadlock around memory overcommit tracker and memory allocation for logging. And it was also fixed, probably I merged fix into this branch.

KochetovNicolai commented on src/Processors/Executors/ExecutorTasks.cpp:
```diff
@@ -32,14 +32,17 @@ void ExecutorTasks::tryWakeUpAnyOtherThreadWithTasks(ExecutionThreadContext & se
     if (!task_queue.empty() && !threads_queue.empty() && !finished)
     {
-        size_t next_thread = self.thread_number + 1 == num_threads ? 0 : (self.thread_number + 1);
+        size_t next_thread = (self.thread_number + 1) % use_threads;
```
I would be happier to avoid the division (especially if for some reason use_threads = 0).
But probably it's ok

KochetovNicolai commented on src/Processors/Executors/PipelineExecutor.cpp:
```diff
@@ -267,70 +274,98 @@ void PipelineExecutor::initializeExecution(size_t num_threads)
+
+        /// Count of threads in use should be updated for proper finish() condition.
+        /// NOTE: this will not decrease `use_threads` below initially granted count
+        tasks.upscale(thread_num + 1);
```
I think we can just make a method `tasks.upscale()` without arguments which just increases `use_threads` under mutex. Why do we need some extra argument? :)

serxa replied:
It is due to the fact the init()/fill() must use use_threads that is equal to initially granted amount of threads to distribute tasks properly, but spawnThreads() is used even for those initially granted threads

serxa committed:
review fix
